File: java/11.java
// 11. Container With Most Water
//     Given n non-negative integers a1, a2, ..., an , where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis forms a container, such that the container contains the most water.
//
//     Note: You may not slant the container and n is at least 2.
//
// Example:
//     Input: [1,8,6,2,5,4,8,3,7]
//     Output: 49

// Runtime: 425 ms, faster than 8.33% of Java online submissions for Container With Most Water.
// Memory Usage: 39.9 MB, less than 94.87% of Java online submissions for Container With Most Water.

class Solution {
    public int maxArea(int[] height) {
        int x1 = 0;
        int x2 = 1;
        int cur = 0;
        int max = 0;
        int fin = height.length;
        while (x2 < fin) {
            cur = (x2 - x1) * Math.min(height[x1], height[x2]);
            if (cur > max) {
                max = cur;
            }
            if (x2 == fin - 1) {
                x1 = x1 + 1;
                x2 = x1 + 1;
            } else {
                x2 = x2 + 1;
            }
        }
        return max;
    }
}


// Runtime: 1 ms, faster than 100.00% of Java online submissions for Container With Most Water.
// Memory Usage: 40 MB, less than 94.87% of Java online submissions for Container With Most Water.
class Solution {
   public int maxArea(int[] height) {
        int x1 = 0;
        int x2 = height.length - 1;
        int max = 0;
        while (x2 != x1) {
            if (height[x1] > height[x2]) {
                max = Math.max((x2 - x1) * height[x2], max);
                x2--;
            } else {
                max = Math.max((x2 - x1) * height[x1], max);
                x1++;
            }
        }
        
        return max;
    }
}

File: go/11.go
// 11. Container With Most Water

// You are given an integer array height of length n. There are n vertical lines drawn such that the two endpoints of the ith line are (i, 0) and (i, height[i]).

// Find two lines that together with the x-axis form a container, such that the container contains the most water.

// Return the maximum amount of water a container can store.

// Notice that you may not slant the container.

 

// Example 1:


// Input: height = [1,8,6,2,5,4,8,3,7]
// Output: 49
// Explanation: The above vertical lines are represented by array [1,8,6,2,5,4,8,3,7]. In this case, the max area of water (blue section) the container can contain is 49.
// Example 2:

// Input: height = [1,1]
// Output: 1
 

// Constraints:

// n == height.length
// 2 <= n <= 105
// 0 <= height[i] <= 104

// Runtime 0 ms Beats 100.00%
// Memory 11.85 MB Beats 10.00%
func maxArea(height []int) int {
    left := 0
    right := len(height) - 1
    max := 0
    for ; left < right; {
        area := 0
        if height[left] < height[right] {
            area = (right - left) * height[left]
            left++
        } else {
            area = (right - left) * height[right]
            right--
        }
        if area > max {
            max = area
        }
    }
    return max
}
